File: java/src/main/java/Direction.java
import java.util.List;
import java.util.ArrayList;

public class Direction {
    public static final int NONE = -1;
    private int direction;

    public Direction() {
        direction = NONE;
    }

    public int getDirection() {
        return direction;
    }

    public void setDirection(String value) {
        
        if (value.equals("N")) {
            direction = 0;
        } 
        else if (value.equals("E")) {
            direction = 1;
        } 
        else if (value.equals("S")) {
            direction = 2;
        } 
        else if (value.equals("W")) {
            direction = 3;
        } 
        else {
            throw new IllegalArgumentException("Invalid Direction: " + value);
        }
    }

    public void setDirection(int value) {
        if (value >= 0 && value <= 3) {
            direction = value;
        } 
        else {
            throw new IllegalArgumentException("Invalid Direction: " + value);
        }
    }

    public List<Integer> getTurns(int dx, int dy) 
    {
        List<Integer> toTurns = new ArrayList<>();
        
        if (isPositive(dx)) {
            toTurns.add(1);
        } 
        else if (isNegative(dx)) {
            toTurns.add(3);
        }
        
        if (isPositive(dy)) {
            toTurns.add(0);
        } 
        else if (isNegative(dy)) {
            toTurns.add(2);
        }
        
        return toTurns;
    }

    public int turnsTaken(List<Integer> changes) {
        int start = direction;
        int count = 0;
        final int MAX_LEN = 4;

        for (int i : changes) {
            count += Math.min((i - start + MAX_LEN) % MAX_LEN, (start - i + MAX_LEN) % MAX_LEN);
            start = i;
        }
        return count;
    }

    public int calcTurns(int dx, int dy) {
        List<Integer> turns = getTurns(dx, dy);

        int ans = Math.min(turnsTaken(turns), turnsTaken(reverseList(turns)));

        return ans;
    }

    private boolean isPositive(int number) {
        return number > 0;
    }

    private boolean isNegative(int number) {
        return number < 0;
    }

    private List<Integer> reverseList(List<Integer> list) {
        List<Integer> reversed = new ArrayList<>(list);
        java.util.Collections.reverse(reversed);
        return reversed;
    }
}

File: cpp/src/direction.cpp
#include "direction.h"

Direction::Direction() : _direction(NONE) {}

int Direction::getDirection() const {
    return _direction;
}

void Direction::setDirection(const std::string& value) {
    if (DB.find(value) != DB.end()) {
        _direction = DB.at(value);
    } else {
        throw std::invalid_argument("Invalid Direction: " + value);
    }
}

void Direction::setDirection(int value) {
    if (value >= 0 && value <= 3) {
        _direction = value;
    } else {
        throw std::invalid_argument("Invalid Direction: " + std::to_string(value));
    }
}

std::vector<int> Direction::getTurns(int dx, int dy) const {
    std::vector<int> toTurns;

    if (dx > 0) {
        toTurns.push_back(DB.at("E"));
    } else if (dx < 0) {
        toTurns.push_back(DB.at("W"));
    }

    if (dy > 0) {
        toTurns.push_back(DB.at("N"));
    } else if (dy < 0) {
        toTurns.push_back(DB.at("S"));
    }

    return toTurns;
}

int Direction::turnsTaken(const std::vector<int>& changes) const {
    int start = _direction;
    int countr = 0;
    int MAX_LEN = DB.size();

    for (int i : changes) {
        countr += std::min((i - start + MAX_LEN) % MAX_LEN, (start - i + MAX_LEN) % MAX_LEN);
        start = i;
    }
    return countr;
}

int Direction::calcTurns(int dx, int dy) const {
    std::vector<int> turns = getTurns(dx, dy);
    int ans = std::min(turnsTaken(turns), turnsTaken(std::vector<int>(turns.rbegin(), turns.rend())));
    return ans;
}
